### PR TITLE
create workflow project from REST API instead of GQL

### PR DESCRIPTION
**Pull request by jhoetter**
This PR converts the GraphQL call to create a refinery project from a GraphQL mutation into a REST endpoint, which is then called from the workflow-backend instead of the workflow-frontend

### CODE REVIEW

**Comment by jhoetter** [comment]
I also added a bugfix for the `import_json` REST API endpoint

File: api/project.py [hunk 1]
```diff
@@ -7,6 +7,29 @@
...
+from sqlalchemy.orm import sessionmaker
+
+__engine = None
+
+
+def get_workflow_db_engine():
+    global __engine
+    if __engine is None:
+        __engine = create_engine(os.getenv("WORKFLOW_POSTGRES"))
+    return __engine
```
Comment (SirDegraf) [inline api/project.py]: database connection/engine management seems not optimal placed in an api file, i would expect some sort of layer for this or at least an util file?

File: api/project.py [hunk 2]
```diff
@@ -41,3 +64,62 @@ def get(self, request) -> JSONResponse:
             "knowledge_base_ids": [str(list.id) for list in project.knowledge_bases],
         }
         return JSONResponse(result)
+
+
+class ProjectCreationFromWorkflow(HTTPEndpoint):
```
Comment (SirDegraf) [inline api/project.py:50]: much of what happens here would be outsourced to another layer in our other services. Request managment (reading data out of request and auth of user) can of course be done in an api, but application logic like checking and creating a bucket or session management of a database could be put into a controller file, at least thats what i would have expected

File: api/project.py [hunk 3]
```diff
@@ -41,3 +64,62 @@ def get(self, request) -> JSONResponse:
...
+        project = project_manager.create_project(
+            organization.id, name, description, user.id
+        )
+        project_manager.update_project(project_id=project.id, tokenizer=tokenizer)
+
+        Session = sessionmaker(get_workflow_db_engine())
+        with Session() as session:
+            results = session.execute(
+                f"SELECT record FROM store_entry WHERE store_id = '{store_id}'"
+            ).all()
```
Comment (SirDegraf) [inline api/project.py]: SQL-Code and its execution should be put in the model submodule

File: api/project.py [hunk 4]
```diff
@@ -41,3 +64,62 @@ def get(self, request) -> JSONResponse:
...
+
+        data = [result for result, in results]
```
Comment (SirDegraf) [inline api/project.py]: maybe replace var names by more expressive naming

File: api/project.py [hunk 5]
```diff
@@ -41,3 +64,62 @@ def get(self, request) -> JSONResponse:
...
+
+        upload_task = upload_task_manager.create_upload_task(
+            user_id=user.id,
+            project_id=project.id,
+            file_name=name,
+            file_type="json",
```
Comment (SirDegraf) [inline api/project.py]: use of enum is recommended for constant strings

File: api/transfer.py [hunk 6]
```diff
@@ -132,10 +137,26 @@ async def post(self, request) -> JSONResponse:
         request_body = await request.json()
```
Comment (SirDegraf) [inline api/transfer.py:137]: again complex application logic in an api

File: api/transfer.py [hunk 7]
```diff
@@ -132,10 +137,26 @@ async def post(self, request) -> JSONResponse:
...
+        for attribute in project.attributes:
+            if attribute.is_primary_key:
+                for idx, record in enumerate(records):
+                    if attribute.name not in record:
+                        if attribute.name == "running_id":
+                            records[idx][attribute.name] = num_project_records + idx + 1
+                        else:
+                            raise exceptions.InvalidInputException(
+                                f"Non-running-id, primary key {attribute.name} missing in record"
+                            )
```
Comment (SirDegraf) [inline api/transfer.py:154]: Are missing fields allowed when they are not primary keys? I think we forbid it for the "normal" upload but maybe this is a wrong information on my side

File: app.py [hunk 8]
```diff
@@ -40,6 +40,7 @@
     Route("/project/{project_id:str}/import_file", PrepareFileImport),
     Route("/project/{project_id:str}/import_json", JSONImport),
     Route("/project/{project_id:str}/import/task/{task_id:str}", UploadTask),
+    Route("/project", ProjectCreationFromWorkflow),  # create fro workflow
```
Comment (SirDegraf) [inline app.py]:
```suggestion
    Route("/project", ProjectCreationFromWorkflow),  # create from workflow
```

Comment (SirDegraf) [reply app.py]: But since the class is already expressive named, i would remove the comment i think

**Comment by SirDegraf** [comment]
For me it seems like something was changed in the submodule model, too. Maybe I am just missing something, but I would expect a linked submodule PR, but I can't find one. Is there one?

image

File: api/project.py (hunk 1 above)
Comment (jhoetter) [reply api/project.py]: created an adapter file in utils for this

File: api/project.py (hunk 3 above)
Comment (jhoetter) [reply api/project.py]: this is SQL to access the workflow db, so i will also put it into the adapter

File: api/project.py (hunk 4 above)
Comment (jhoetter) [reply api/project.py]: good advice, changed it to:
```python
def get_records_from_store(store_id: str) -> List[Dict[str, Any]]:
    Session = sessionmaker(get_workflow_db_engine())
    with Session() as session:
        record_entity_list = session.execute(
            f"SELECT record FROM store_entry WHERE store_id = '{store_id}'"
        ).all()

    record_dict_list = [result for result, in record_entity_list]
    return record_dict_list
```

File: api/project.py (hunk 5 above)
Comment (jhoetter) [reply api/project.py]: added these enums:
```python
class RecordImportFileTypes(Enum):
    JSON = "JSON"
    CSV = "CSV"
    XLSX = "XLSX"

```

File: api/transfer.py (hunk 7 above)
Comment (jhoetter) [reply api/transfer.py:154]: They should be allowed, just considered as NaN fields. I think this is already happening in refinery, right?

File: app.py (hunk 8 above)
Comment (jhoetter) [reply app.py]: makes sense, removed it :)

**Comment by jhoetter** [comment]
> For me it seems like something was changed in the submodule model, too. Maybe I am just missing something, but I would expect a linked submodule PR, but I can't find one. Is there one?
> 
> image

yes, that was a mistake on my side, didn't push the submodules. I always forget about them...

File: api/project.py (hunk 2 above)
Comment (jhoetter) [reply api/project.py:50]: where applicable, i transferred it into the controller logic of project (`add_workflow_store_data_to_project`)

File: api/transfer.py (hunk 6 above)
Comment (jhoetter) [reply api/transfer.py:137]: created a method `unpack_request_body` in the adapter for this

**Review by SirDegraf — APPROVED** [review]

**Pull request merged**